malloc/src/mymalloc.c

```diff
@@ -4,19 +4,80 @@
 
 #define ALIGNMENT 8 // sizeof(double)
 #define ALIGN(size) (((size) + (ALIGNMENT - 1)) & ~(ALIGNMENT))
-#define HEADER_SIZE sizeof(header_t)
+#define HEADER_SIZE sizeof(header_t)/sizeof(double) + (sizeof(header_t)%sizeof(double) != 0) //size of header in blocks of 8 bytes
+#define HEAP_SIZE 512
+#define END &heap[HEAP_SIZE]  //one past the end of the heap
+
+static double heap[HEAP_SIZE+1]; 
 
 typedef struct memory_header {
-    int size;
+    int size; 
     bool isFree;
 } header_t;
 
+//given a block header, return the next block, or ptr to one past heap if it does not exist
+header_t* nextHeader(header_t *header){
+    return (header_t *) &header[header->size + HEADER_SIZE];
+}
+
+//prints current state of memory, for debugging perposes
+void printMemory(){ 
+    header_t *header = (header_t *) &heap[0];
+    while (header != END) {
+        printf("size: %d, isFree: %d\n", header->size*8, header->isFree);
+        header = nextHeader(header);
+    }
+}
+
+//given a header block of free memory, cut it to desired size leaving an empty suffix
+//this function assumes that header->size >= size
+//if thats not true *BAD* things will happen
+header_t* cut(header_t *header, int size){
+    if(header->size == size){
+        header->isFree = false;
+        return header;
+    }
+    header_t *newHeader = (header_t *) &header[size + HEADER_SIZE];
+    newHeader->size = header->size - size - HEADER_SIZE;
+    newHeader->isFree = true;
+
+    header->size = size;
+    header->isFree = false;
+    return header;
+}
+
 void *mymalloc(size_t size, char *file, int line) {
-    printf("Function called from %s, on line number: %d", file, line);
-    int dummyAddress = 5;
-    return &dummyAddress;
+    //turn size into # of blocks of 8 bytes needed
+    size = (size/8) + ((size%8) != 0);
```

Comment:
hmm, maybe might be better to keep size as specified by the user, as it is more correct. i can do this implementation if needed, but I think it would just be an annoying # of ALIGN() macro calls.
btw `ALIGN()` does the exact thing as this.